stdlib/public/runtime/ObjCRuntimeGetImageNameFromClass.mm

```diff
@@ -238,11 +260,62 @@ static const char *patchedGetImageNameFromClassForOldOSs(Class _Nullable cls) {
   if (!cls)
     return nullptr;
   const char *result;
-  if (getImageNameFromSwiftClass(cls, &result))
+  if (replacementGetImageNameFromClass(cls, &result))
     return result;
   return nullptr;
 }
 
+#if PATCH_NSBUNDLE
+/// Selectors for the target method, patch method, and helper method.
+#define BUNDLE_FOR_CLASS_SEL @selector(bundleForClass:)
+#define PATCHED_BUNDLE_FOR_CLASS_SEL @selector(_swift_bundleForClass:)
+#define BUNDLE_WITH_EXECUTABLE_PATH_SEL @selector(_swift_bundleWithExecutablePath:)
```

Comment:
[GitHub ate this comment the first time I submitted it; trying again.]

It seems weird to me to define a normal method for `_swift_bundleWithExecutablePath:` but patch `_swift_bundleForClass:` into place with `class_addMethod`. Why not either

- always provide `+_swift_bundleForClass:`, and collapse `+_swift_bundleWithExecutablePath:` into it (but only swap them conditionally)
- collapse `+_swift_bundleWithExecutablePath:` into `patchedBundleForClass`, since it doesn't actually use `self`

?

You can also avoid some of the `method_exchangeImplementations` dance with `method_setImplementation`, saving the old IMP in a static property instead of another method.

Reply:
I organized it this way to keep the dependencies clean(er). The runtime side shouldn't link anything in from Foundation, but `_swift_bundleForClass:` wants to use the `getImageNameFromSwiftClass` in the runtime. I could do everything in `patchedBundleForClass` but it would be inconvenient to look up all of that stuff at runtime.

I used `method_exchangeImplementations` for thread safety. `method_setImplementation` would have a race condition if another thread was calling `+bundleForClass:` when the patch was being installed, since there's a window where the patch is active but the old `IMP` hasn't yet been stored into the `static`. Could work around that by fetching it first but then there's potential (albeit unlikely) races with other swizzlers.

Reply:
Ooh, good catch. Okay, thanks for the explanation.

Reply:
I was this close to committing a version with `method_setImplementation` when I had a funny thought in the back of my mind about thread safety. I've always done swizzling like that in the past, but I've always done it at startup rather than lazily like this one.